### PR TITLE
ensure that terminationRunnable is invoked only once during shutdown

### PR DESCRIPTION
**Pull request by dconnelly**

### CODE REVIEW

**Comment by dconnelly** [comment]
fixes #517

**Comment by ejona86** [comment]
Although this would fix #517, it also would cause us to be considered terminated prematurely. If we don't wait until the transportServer terminates, then the server port may still be open and we could still be using some executors.

**Comment by dconnelly** [comment]
Not sure I follow. State 'terminated' is still set under the same conditions as before, once all transports have been closed and we have entered the shutdown state. This change just makes sure we don't run through the termination cleanup process again once we've already reached the terminated state.

**Comment by ejona86** [comment]
The problem is that we shouldn't have run through the termination cleanup process the first time. We only want to run it the second time. The first time is actually premature.

**Comment by dconnelly** [comment]
Ok, I see now. Sorry I had confused transportServer with serverTransport :) I'll fix and issue a new pull request.

**Comment by dconnelly** [comment]
Eric, I added a check for transportServer termination to the condition. Let me know if this does the trick.

File: core/src/main/java/io/grpc/ServerImpl.java [hunk 1]
```diff
@@ -220,6 +221,7 @@ public void serverShutdown() {
       synchronized (ServerImpl.this) {
         // transports collection can be modified during shutdown(), even if we hold the lock, due
         // to reentrancy.
+        transportServerTerminated = true;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ServerImpl.java]: If we move this assignment to after the for loop (and just before `checkForTermination()`) then it seems that `!terminated` would always be true (and so we wouldn't need that in `checkForTermination()`'s condition).

At that point we could make !terminated an assert (either with the `assert` keyword or `if (!terminated) throw new AssertionError()` to make sure we notice if something breaks in the future.

**Comment by ejona86** [comment]
@dconnelly, awesome. LGTM and thanks!

Cherrypicked as 15104cd

**Pull request closed without merging**